lib/traject/macros/stanford.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+module Macros
+  # Macros for extracting Stanford Specific MODS values from Nokogiri documents
+  module Stanford
+    NS = { mods: 'http://www.loc.gov/mods/v3',
+           rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+           dc: 'http://purl.org/dc/elements/1.1/' }.freeze
+
+    def generate_druid(record, context)
+      if context.settings.fetch('identifier')
+        identifier = context.settings.fetch('identifier').dup
+        identifier.sub! 'stanford_', ''
+      else
+        record.xpath('/*/mods:identifier', NS).map(&:text)
+      end
+    end
+
+    def druid?(identifier)
+      identifier =~ /([a-z]{2})(\d{3})([a-z]{2})(\d{4})\z/
+    end
+
+    def generate_sul_shown_at
+      lambda do |record, accumulator, context|
+        druid = generate_druid(record, context)
+        mods_url = record.xpath('/*/mods:location/mods:url', NS).map(&:text)
+        accumulator.concat(mods_url) if mods_url.present?
+        accumulator << "https://purl.stanford.edu/#{druid}"
+      end
+    end
+
+    def grab_sul_iiif_links(manifest)
+      result = Net::HTTP.get(URI.parse(manifest))
+      process_iiif_json(result)
+    rescue
+      nil
+    end
+
+    def process_iiif_json(result)
+      resp = JSON.parse(result)
+      iiif_service(resp['sequences'])
+      generate_sul_preview(resp['thumbnail']['@id'])
+      iiif_protocol(resp['thumbnail']['service']['profile'])
+      iiif_service_conforms_to(resp['thumbnail']['service']['profile'])
+    rescue JSON::ParserError
+      nil
+    end
+
+    def generate_sul_preview(thumbnail_uri)
+      to_field 'agg_preview', literal(thumbnail_uri)
+    end
+
+    def iiif_service(sequences)
+      to_field 'wr_has_service' do |accumulator|
+        sequences.each do |sequence|
+          sequence['canvases'].each do |canvas|
+            canvas['images'].each do |image|
```

Comment:
Yep, and the application profile suggests that `agg_is_shown_by` should have a cardinality of 0..1.